designs/1951-required-fields-v2.md

```diff
@@ -0,0 +1,352 @@
+
+1. A field is defined with the given value, which must evaluate without error (regular field)
+1. A field _may_ be defined, in which case given constraints apply (optional field).
+1. A field _must_ be defined elsewhere (required field).
+1. A field must not exist (optional field constraint where the constraint is an error value).
+1. A field is only defined if some condition is met (conditional field).
+
+In effect there are two types of fields: those that provide constraints and those that populate values. Below we discuss these use cases within particular usage contexts.
+
+### Data and export
+In CUE, we classify data as [regular fields](https://cuelang.org/docs/references/spec/#definitions-and-hidden-fields)
+(so no definitions or hidden fields), with only concrete values.
+In a nutshell, any JSON value is data in CUE.
+
+In contrast to JSON, but like YAML, CUE permits fields with top-level references to other data fields
+to be classified as data.
+
+### Schema and constraints
+There are two common cases for defining fields in schema:
+
+1. If the user specifies a field `x`, apply constraint `c`.
+1. The user _must_ specify a field `x` with constraint `c`.
+
+Orthogonally to this, the user may wish to specify that `c` must be concrete.
+This is not done explicitly, but rather is enforced implicitly in data mode
+by requiring that any regular field (so not a constraint `a?: x`) must be concrete.
+
+Case 1 can be expressed as:
+```
+x? : c
+```
+Case 2 is currently expressed as:
+```
+x : c
+```
+The requirement that the user must specify a field is only implicit.
+This works as desired when `c` is not concrete: because “non-optional” fields are required to be concrete,
+it will force the user to specify the field with the concrete value.
+
+It is not possible, however, to force a user to specify a field with a specific concrete value.
+This includes any list or struct.
+This is because there is no syntactic distinction between a field that is a constraint,
+but required, and a regular field: the required property is implied by the fact that regular
+fields must be concrete upon output.
+That is, a regular field containing a concrete value that must be specified is
+indistinguishable from a regular field with a concrete value specified by the user.
+
+See, for instance, https://github.com/cue-lang/cue/issues/740 where this is problematic.
+
+### Derived fields and templating
+The purpose of derived fields and templating is to automatically populate fields for a user.
+Possible use cases are:
+
+* providing a set of defaults, 
+* defining a higher level of abstraction that is mapped to some more concrete representation, or 
+* converting between two versions of an API.
+
+Examples of such fields are:
+
+1. Defaults: `a: *1 | int`
+1. Derived values: `a: b + c`
+1. Derived constraints: `a: b + c`, but only if `b + c` is defined.
+
+All derived fields use regular, non-optional, fields.
```

Comment:
This is an assertion that's not necessarily true (it's perfectly feasible to define a required or optional field as a derived field AFAICS). Maybe "should use" ? I'm not even sure that that's true tbh.

Reply:
Good point. That is absolutely true.

I'll removed that sentence.